### PR TITLE
add support for aesara objectives

### PR DESCRIPTION
**Pull request by FFroehlich**

### CODE REVIEW

File: pypesto/objective/aesara.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        def _log_prob_hess(x):
+            return - beta * self._objective(x, sensi_orders=(2,))
+
+        self._log_prob_hess = _log_prob_hess
+
+    def perform(self, node, inputs, outputs, params=None):
+        theta, = inputs
+        # calculate gradients
+        log_prob_grad = self._log_prob_hess(theta)
+        outputs[0][0] = log_prob_grad
```
Comment (yannikschaelte) [inline pypesto/objective/aesara.py:142]:
```suggestion
        log_prob_hess = self._log_prob_hess(theta)
        outputs[0][0] = log_prob_hess
```

File: pypesto/objective/aesara.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        self._objective: Objective = problem.objective
+
+        def _log_prob_hess(x):
+            return - beta * self._objective(x, sensi_orders=(2,))
+
+        self._log_prob_hess = _log_prob_hess
+
+    def perform(self, node, inputs, outputs, params=None):
+        theta, = inputs
+        # calculate gradients
```
Comment (yannikschaelte) [inline pypesto/objective/aesara.py:140]:
```suggestion
        # calculate Hessian
```

File: test/base/test_objective.py [hunk 3]
```diff
@@ -160,3 +165,40 @@ def rel_err(eps_):
         objective.check_grad_multi_eps([theta], multi_eps=multi_eps)
     assert result_multi_eps['rel_err'].squeeze() == \
         min(rel_err(_eps) for _eps in multi_eps)
+
+
+def test_aesara(max_sensi_order, integrated):
```
Comment (yannikschaelte) [inline test/base/test_objective.py:170]: docstring

File: test/base/test_objective.py [hunk 4]
```diff
@@ -160,3 +165,40 @@ def rel_err(eps_):
...
+    assert obj(prob['x']) == prob['obj'](np.sinh(prob['x']))
+
+    if max_sensi_order > 0:
+        df = obj.check_grad(prob['x'])
+        assert (df.abs_err < 1e-6).all() & (df.rel_err < 1e-6).all()
+
+    if max_sensi_order > 1:
+        hess = obj(prob['x'], sensi_orders=(2,))
+        assert np.isclose(hess[0, 0], -468.08047745752054)
+        assert np.isclose(hess[1, 1], 752.4391382167262)
```
Comment (yannikschaelte) [inline test/base/test_objective.py:203]: an analytic expression would be more credible, but should be fine ;)

File: test/base/test_objective.py [hunk 5]
```diff
@@ -160,3 +165,40 @@ def rel_err(eps_):
...
+def test_aesara(max_sensi_order, integrated):
+    prob = rosen_for_sensi(max_sensi_order,
+                           integrated, [0, 1])
+
+    base_objective = AesaraLogProbability(pypesto.Problem(
+        prob['obj'], -np.inf * np.ones_like(prob['x']),
+        +np.inf * np.ones_like(prob['x'])),
+        beta=-1
+    )
+    x = aet.specify_shape(aet.vector('x'), (2,))
```
Comment (yannikschaelte) [inline test/base/test_objective.py:179]: some comments what's going on here?

**Review by yannikschaelte — APPROVED** [review]
Assuming aesara works internally much like theano, looks good. Could you shortly say what the usecase for aesara actually is?

**Comment by FFroehlich** [comment]
> Assuming aesara works internally much like theano, looks good. Could you shortly say what the usecase for aesara actually is?

Aesara is the new pymc maintained Theano fork. Usecase is that it allows composition of objective functions with additional preprocessing where all derivatives are computed automatically via backpropagation.

**Comment by yannikschaelte** [comment]
> > Assuming aesara works internally much like theano, looks good. Could you shortly say what the usecase for aesara actually is?
> 
> Aesara is the new pymc maintained Theano fork. Usecase is that it allows composition of objective functions with additional preprocessing where all derivatives are computed automatically via backpropagation.

Ah sure :facepalm: didn't remember what pymc-theano was called.

File: test/base/test_objective.py (hunk 3 above)
Comment (FFroehlich) [reply test/base/test_objective.py:170]: there is a docstring? arguments come from fixtures

File: test/base/test_objective.py (hunk 4 above)
Comment (FFroehlich) [reply test/base/test_objective.py]: added analytical solutions

File: test/base/test_objective.py (hunk 5 above)
Comment (FFroehlich) [reply test/base/test_objective.py:181]: added

**Pull request merged**